File: bindings/lua/Polycode/Rectangle.lua
class "Rectangle"


function Rectangle:__getvar(name)
	if name == "x" then
		return Polycode.Rectangle_get_x(self.__ptr)
	elseif name == "y" then
		return Polycode.Rectangle_get_y(self.__ptr)
	elseif name == "w" then
		return Polycode.Rectangle_get_w(self.__ptr)
	elseif name == "h" then
		return Polycode.Rectangle_get_h(self.__ptr)
	end
end

function Rectangle:__setvar(name,value)
	if name == "x" then
		Polycode.Rectangle_set_x(self.__ptr, value)
		return true
	elseif name == "y" then
		Polycode.Rectangle_set_y(self.__ptr, value)
		return true
	elseif name == "w" then
		Polycode.Rectangle_set_w(self.__ptr, value)
		return true
	elseif name == "h" then
		Polycode.Rectangle_set_h(self.__ptr, value)
		return true
	end
	return false
end
function Rectangle:Rectangle(...)
	local arg = {...}
	for k,v in pairs(arg) do
		if type(v) == "table" then
			if v.__ptr ~= nil then
				arg[k] = v.__ptr
			end
		end
	end
	if self.__ptr == nil and arg[1] ~= "__skip_ptr__" then
		self.__ptr = Polycode.Rectangle(unpack(arg))
	end
end

function Rectangle:setRect(x, y, w, h)
	local retVal = Polycode.Rectangle_setRect(self.__ptr, x, y, w, h)
end

function Rectangle:Clipped(rect)
	local retVal = Polycode.Rectangle_Clipped(self.__ptr, rect.__ptr)
	if retVal == nil then return nil end
	local __c = _G["Rectangle"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function Rectangle:minX()
	local retVal =  Polycode.Rectangle_minX(self.__ptr)
	return retVal
end

function Rectangle:maxX()
	local retVal =  Polycode.Rectangle_maxX(self.__ptr)
	return retVal
end

function Rectangle:minY()
	local retVal =  Polycode.Rectangle_minY(self.__ptr)
	return retVal
end

function Rectangle:maxY()
	local retVal =  Polycode.Rectangle_maxY(self.__ptr)
	return retVal
end

function Rectangle:__delete()
	if self then Polycode.delete_Rectangle(self.__ptr) end
end

File: bindings/javascript/Polycode/Rectangle.js
function Rectangle() {
	if(arguments[0] != "__skip_ptr__") {
		this.__ptr = Polycode.Rectangle()
	}
	Object.defineProperties(this, {
		'x': { enumerable: true, configurable: true, get: Rectangle.prototype.__get_x, set: Rectangle.prototype.__set_x},
		'y': { enumerable: true, configurable: true, get: Rectangle.prototype.__get_y, set: Rectangle.prototype.__set_y},
		'w': { enumerable: true, configurable: true, get: Rectangle.prototype.__get_w, set: Rectangle.prototype.__set_w},
		'h': { enumerable: true, configurable: true, get: Rectangle.prototype.__get_h, set: Rectangle.prototype.__set_h}
	})
}


Rectangle.prototype.__get_x = function() {
	return Polycode.Rectangle__get_x(this.__ptr)
}

Rectangle.prototype.__set_x = function(val) {
	Polycode.Rectangle__set_x(this.__ptr, val)
}

Rectangle.prototype.__get_y = function() {
	return Polycode.Rectangle__get_y(this.__ptr)
}

Rectangle.prototype.__set_y = function(val) {
	Polycode.Rectangle__set_y(this.__ptr, val)
}

Rectangle.prototype.__get_w = function() {
	return Polycode.Rectangle__get_w(this.__ptr)
}

Rectangle.prototype.__set_w = function(val) {
	Polycode.Rectangle__set_w(this.__ptr, val)
}

Rectangle.prototype.__get_h = function() {
	return Polycode.Rectangle__get_h(this.__ptr)
}

Rectangle.prototype.__set_h = function(val) {
	Polycode.Rectangle__set_h(this.__ptr, val)
}

Duktape.fin(Rectangle.prototype, function (x) {
	if (x === Rectangle.prototype) {
		return;
	}
	Polycode.Rectangle__delete(x.__ptr)
})

Rectangle.prototype.setRect = function(x,y,w,h) {
	Polycode.Rectangle_setRect(this.__ptr, x, y, w, h)
}

Rectangle.prototype.Clipped = function(rect) {
	var retVal = new Rectangle("__skip_ptr__")
	retVal.__ptr = Polycode.Rectangle_Clipped(this.__ptr, rect)
	return retVal
}

Rectangle.prototype.minX = function() {
	return Polycode.Rectangle_minX(this.__ptr)
}

Rectangle.prototype.maxX = function() {
	return Polycode.Rectangle_maxX(this.__ptr)
}

Rectangle.prototype.minY = function() {
	return Polycode.Rectangle_minY(this.__ptr)
}

Rectangle.prototype.maxY = function() {
	return Polycode.Rectangle_maxY(this.__ptr)
}
